src/main/java/io/reactivex/internal/operators/single/SingleTimeout.java

```diff
@@ -43,97 +44,116 @@ public SingleTimeout(SingleSource<T> source, long timeout, TimeUnit unit, Schedu
 
-        TimeoutDispose(AtomicBoolean once, CompositeDisposable set, SingleObserver<? super T> s) {
-            this.once = once;
-            this.set = set;
-            this.s = s;
-        }
+        final TimeoutFallbackObserver<T> fallback;
 
-        @Override
-        public void run() {
-            if (once.compareAndSet(false, true)) {
-                if (other != null) {
-                    set.clear();
-                    other.subscribe(new TimeoutObserver());
-                } else {
-                    set.dispose();
-                    s.onError(new TimeoutException());
-                }
-            }
-        }
+        SingleSource<? extends T> other;
 
-        final class TimeoutObserver implements SingleObserver<T> {
+        static final class TimeoutFallbackObserver<T> extends AtomicReference<Disposable>
+        implements SingleObserver<T> {
 
-            @Override
-            public void onError(Throwable e) {
-                set.dispose();
-                s.onError(e);
+            private static final long serialVersionUID = 2071387740092105509L;
+            final SingleObserver<? super T> actual;
+
+            TimeoutFallbackObserver(SingleObserver<? super T> actual) {
+                this.actual = actual;
             }
 
             @Override
             public void onSubscribe(Disposable d) {
-                set.add(d);
+                DisposableHelper.setOnce(this, d);
             }
 
             @Override
-            public void onSuccess(T value) {
-                set.dispose();
-                s.onSuccess(value);
+            public void onSuccess(T t) {
+                actual.onSuccess(t);
             }
 
+            @Override
+            public void onError(Throwable e) {
+                actual.onError(e);
+            }
         }
-    }
 
-    final class TimeoutObserver implements SingleObserver<T> {
+        TimeoutMainObserver(SingleObserver<? super T> actual, SingleSource<? extends T> other) {
+            this.actual = actual;
+            this.other = other;
+            this.task = new AtomicReference<Disposable>();
+            if (other != null) {
+                this.fallback = new TimeoutFallbackObserver<T>(actual);
+            } else {
+                this.fallback = null;
+            }
+        }
 
-        private final AtomicBoolean once;
-        private final CompositeDisposable set;
-        private final SingleObserver<? super T> s;
+        @Override
+        public void run() {
+            Disposable d = get();
+            if (d != DisposableHelper.DISPOSED && compareAndSet(d, DisposableHelper.DISPOSED)) {
+                if (d != null) {
+                    d.dispose();
+                }
+                SingleSource<? extends T> other = this.other;
+                if (other == null) {
+                    actual.onError(new TimeoutException());
+                } else {
+                    this.other = null;
```

Comment:
Hmm, are you trying to avoid memory leak here? Reference to other is still held as a final field in `SingleTimeout`

I mean, final reference to `other` in `TimeoutMainObserver` will make code slightly more readable

Reply:
The outer class has to reference other because of multiple subscribers. The inner observer has no use to the other once it has switched to it. There was an issue that some time ago that other was some cached resource that didn't go away when the user wanted it. This will somewhat help in that situation.